### PR TITLE
Updated README with app level index routing

### PR DESCRIPTION
**Pull request by harshit-splunk**

### CODE REVIEW

**Review by luckyj5 — CHANGES_REQUESTED** [review]
Lets also add that "ADD_APP_INFO" (cache info) is required to be enabled for per app index routing feature. 
If Add-app-info is not enabled, nozzle will not have app's metadata which will include SPLUNK_INDEX value.

File: README.md [hunk 1]
```diff
@@ -158,26 +158,35 @@ specifying correct "--boltdb-path" flag or "BOLTDB_PATH" environment variable.
...
+  ...
   env:
-    SPLUNK_INDEX: testing_index
+    SPLUNK_INDEX: <SPLUNK_INDEX>
+    ...
+```
+
+You can also update the env on the fly using cf-cli command:
+```
+cf set-env <APP_NAME> SPLUNK_INDEX <ENV_VAR_VALUE>
```
Comment (luckyj5) [inline README.md:184]: I think for this, if you set-env via CLI, you still need to restage the application for those updated envs to take effect. Please cross check. Thanks!

**Review by luckyj5 — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (harshit-splunk) [reply README.md:184]: > I think for this, if you set-env via CLI, you still need to restage the application for those updated envs to take effect. Please cross check. Thanks!

In my testing, I was able to route data-gen events to the index provided using cf command. I tested following scenarios:
- deployed data_gen without setting SPLUNK_INDEX env variable and set env after it is deployed
- deployed data_gen with setting SPLUNK_INDEX env variable and update env with new index value
In both cases, event were ingested in newly specified index.

I think, if the app is using the environment variable then we are required to restage the application. For our use case, app doesn't rely on SPLUNK_INDEX, so no need to restage. Cloud controller endpoint will always returns latest env data. So as long as connector is invalidating app cache, it should be fine

**Comment by harshit-splunk** [comment]
> Lets also add that "ADD_APP_INFO" (cache info) is required to be enabled for per app index routing feature.
> If Add-app-info is not enabled, nozzle will not have app's metadata which will include SPLUNK_INDEX value.

Sure

**Comment by luckyj5** [comment]
> > Lets also add that "ADD_APP_INFO" (cache info) is required to be enabled for per app index routing feature.
> > If Add-app-info is not enabled, nozzle will not have app's metadata which will include SPLUNK_INDEX value.
> 
> It doen't depend on ADD_APP_INFO config. As you can see in [router](https://github.com/cloudfoundry-community/splunk-firehose-nozzle/blob/edd2401b7a787a1f59c19d619f07c4a258b11c68/eventrouter/default.go#L70), as long as event has app_id, it will call event.AnnotateWithAppData method. In event.AnnotateWithAppData method, it gets app information from cache from cf_app_id. If it find's SPLUNK_INDEX in app env, it'll add _info_splunk_index_ field in the event

ADD_APP_INFO config triggers nozzle to fetch cache info for apps. If it's not selected, then nozzle won't have cache info, which means it won't have env variables either. Let me know if it's otherwise.

**Comment by harshit-splunk** [comment]
> > > Lets also add that "ADD_APP_INFO" (cache info) is required to be enabled for per app index routing feature.
> > > If Add-app-info is not enabled, nozzle will not have app's metadata which will include SPLUNK_INDEX value.
> > 
> > 
> > It doen't depend on ADD_APP_INFO config. As you can see in [router](https://github.com/cloudfoundry-community/splunk-firehose-nozzle/blob/edd2401b7a787a1f59c19d619f07c4a258b11c68/eventrouter/default.go#L70), as long as event has app_id, it will call event.AnnotateWithAppData method. In event.AnnotateWithAppData method, it gets app information from cache from cf_app_id. If it find's SPLUNK_INDEX in app env, it'll add _info_splunk_index_ field in the event
> 
> ADD_APP_INFO config triggers nozzle to fetch cache info for apps. If it's not selected, then nozzle won't have cache info, which means it won't have env variables either. Let me know if it's otherwise.

On further analysis, found that boltdb cache is initialised only if ADD_APP_INFO is enabled otherwise NoCache is initialised. And app info will never be fetched from remote.
I'll make changes according to that.

**Review by kashyap-splunk — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -157,26 +157,35 @@ specifying correct "--boltdb-path" flag or "BOLTDB_PATH" environment variable.
 Set F2S_DISABLE_LOGGING = true as a environment variable in applications's manifest to disable logging.
 
 
-### Index routing
+## Index routing
 Index routing is a feature that can be used to send different Cloud Foundry logs to different indexes for better ACL and data retention control in Splunk.
 
-#### Per application index routing via application manifest
+### Per application index routing via application manifest
+To enable per app index routing, you are required to set environmt variable `SPLUNK_INDEX`in your application and enable ADD_APP_INFO(Select at least one metadata) to enable app info caching. Make sure that given HEC token can ingest events to the SPLUNK_INDEX.
```
Comment (luckyj5) [inline README.md:164]: Please update as below: 

To enable per application index routing:
- Please set environment variable `SPLUNK_INDEX` in your application's manifest (example below - link to the example) 
- Make sure Splunk nozzle is configured with ADD_APP_INFO (Select at least one of AppName, OrgName/Id, SpaceName/id) to enable app info caching
- Make sure SPLUNK_INDEX specified in app's manifest exist in Splunk and can receive data for the configured Splunk HEC token.

**Comment by luckyj5** [comment]
Thanks for update @harshit-splunk. 
Just noticed this - [note](https://github.com/cloudfoundry-community/splunk-firehose-nozzle#please-note)
Please remove this as well.

**Review by luckyj5 — CHANGES_REQUESTED** [review]
Just noticed this - [note](https://github.com/cloudfoundry-community/splunk-firehose-nozzle#please-note) for SPLUNK_VERSION
Please remove this as well.

**Review by luckyj5 — APPROVED** [review]
LGTM

**Pull request merged**